[maze508] iP

maze508 opened the pull request:
# Duke Task Manager :rocket:

> "Your mind is for having ideas, not holding them." – David Allen ([source](https://www.goodreads.com/quotes/348103-your-mind-is-for-having-ideas-not-holding-them/))

**Duke Task Manager** helps you keep track of your tasks effortlessly, freeing your mind for more important things. It's:

- *text-based*
- **easy to learn**
- ~~slow~~ **SUPER FAST** to use

## Getting Started

To use Duke Task Manager, follow these steps:

1. Download it from [here](https://example.com/duke.jar).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉
5. Enjoy being productive!

And the best part? It's **FREE**!

## Features

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

### For Java Programmers

If you're looking to practice Java, Duke Task Manager can be a good starting point. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

maze508 committed:
level 0

maze508 committed:
Level-1: Echo

maze508 committed:
Level-2: Add, List

maze508 committed:
Level-3: Mark as Done

maze508 committed:
Level-4: ToDos, Events, Deadlines

maze508 committed:
A-TextUiTesting

maze508 committed:
Level-5: Handle Errors

maze508 committed:
Level-6: Delete

maze508 committed:
Added logic for Level-7 Safe

maze508 committed:
Add dates and times logic

maze508 committed:
Modularise implementation for A-MoreOOP

maze508 committed:
Fix file creation upon startup

maze508 committed:
Organise into Packages

maze508 committed:
Merge branch 'master' into add-gradle-support

maze508 committed:
Fix application class name

maze508 committed:
Remove storage txt

maze508 committed:
Add Unit Tests for TaskList, Todo, Event & Deadline

maze508 committed:
Update build.gradle

maze508 committed:
Add javadocs for all methods

maze508 committed:
Fix if statement and switch case statements

maze508 committed:
Level-9: Find

maze508 committed:
add config files

maze508 committed:
Fix Styling Issues

maze508 committed:
Edit Javadocs

maze508 committed:
Complete Tutorial

maze508 committed:
Remove duke.txt

maze508 committed:
Update GUI Design

maze508 committed:
Implement Dark Mode in FXML

maze508 committed:
Update Image

maze508 committed:
Add missing comments

wongkj12 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,66 @@
+
+
+/**
+ * The main class for the Duke application.
```
I like the comprehensive javadocs for every class.

wongkj12 reviewed (commented):
LGTM, overall your code is easy to read and I couldn't find any major coding standard issues. Nice images for DaDuke and DaUser :)

maze508 committed:
Fix UI Errors and fixed missing chatting functionality

maze508 committed:
Implement assertions across the project

This commit adds assertion to validate assumptions in the TaskList,
Storage and Duke classes, ensuring that important invariants hold.

Assertions check that:
- No null tasks are added to the TaskList
- The Storage class correctly handles filepath and filepath cannot be
  empty
- Duke ensures that the storage and ui class are well defined

maze508 committed:
Merge pull request #1 from maze508/branch-A-Assertions

Implement assertions across the project

maze508 committed:
Enhance Code Quality

Introduced Assertions and Updated Variable and Function Parameter names
in Parser and Storage to make code more intuitive

maze508 committed:
Merge pull request #2 from maze508/branch-A-CodeQuality

Enhance Code Quality

maze508 committed:
Add Duplicate Detection

maze508 committed:
Merge pull request #3 from maze508/branch-C-DetectDuplicates

Add Duplicate Detection

gowthaman-01 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,151 @@
+        String description = parts[2];
+
+        switch (type) {
+        case "T":
```
Try to use an enum instead of string matching.

tahsinhasem commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,9 @@
+todo borrow book
```
Consider adding testing for erroneous input. We ideally want to handle errors effectively throughout our dev progress.

tahsinhasem commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,71 @@
+package duke;
```
I like the package system. Especially what you did with resources nad testing! However, perhaps consider having Task, Event, Todo, TaskList in one package; Ui code in another package, and a package for utilities containing the Storage, Parser etc. classes.

tahsinhasem commented on src/main/java/duke/DateChecker.java:
```diff
@@ -0,0 +1,38 @@
+package duke;
```
I really Like how succinct and readable the code is.

tahsinhasem commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,15 @@
+package duke;
```
Consider adding other exceptions for handling other cases: EG: DukeWrongCommandExeption, DukeWrongParameterException, DukeInvaliddateException etc. which extend from the DukeException class.

tahsinhasem commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,82 @@
+     * @return Whether the task is equal to the object.
+     */
+    @Override
+    public boolean equals(Object obj) {
```
Nice work with all the `equals` functions!!

tahsinhasem commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,194 @@
+            return ui.showTaskList(tasks);
+        }
+
+        String[] parts = input.split(" ", 2);
```
Good naming choices

tahsinhasem commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,194 @@
+        String commandWord = parts[0];
+        switch (commandWord) {
+            case "todo":
+                return addTodo(parts[1], tasks, ui, storage);
```
making method calls to helper functions instead of having a big block of logic makes the code much more readable!

tahsinhasem commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,194 @@
+     * @return The response message after marking or unmarking the task.
+     * @throws DukeException If the command is invalid or execution fails.
+     */
+    private static String markOrUnmarkTask(String description, TaskList tasks, Ui ui,
```
I am unsure of this convention. Is this the convention we are supposed to follow? Can help me anyone?

tahsinhasem reviewed (commented):
Great work! Code is succint and follows the coding standard of the mod. However, I have left some suggestions here and there!

jovantanyk commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,68 @@
+     * @return The description of the saved task.
+     */
+    @Override
+    public String saveData() {
```
Since the return value is a String, the naming is slightly misleading. getData() or getFileData() would be more accurate

jovantanyk commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Represents a task in the Duke application.
+ */
+public class Task {
```
Since Tasks are never instantiated directly and instead is solely a parent class for Deadline, Todo, and Event, you should make it an abstract class.

jovantanyk commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,151 @@
+/**
+ * Handles loading and saving of tasks to a file.
+ */
+public class Storage {
```
This implementation seems redundant. If FILE_PATH is already a static constant, there is no need to have this.filePath = FILE_PATH. You can remove the entire constructor and only have static methods in Storage since there isn't any reason why we need several instances of Storage

jovantanyk commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,82 @@
+     *
+     * @return The start time of the event.
+     */
+    public LocalDate getStart() {
```
Method names could be slightly more detailed to reflect the return value. getStartDate() and getEndDate() for example.

jovantanyk reviewed (commented):
Overall good code and consistent method names. Task files and Utility files like Parser, Storage and Ui could be separated in different packages. Very good handling of errors and exceptions, and good demonstration of OOP

maze508 committed:
Add Title

maze508 committed:
Update README

maze508 committed:
Add Image to User Guide

maze508 committed:
Fix build files and jar file not being able to be ran

maze508 committed:
Fix empty task not catching issue
